### PR TITLE
C++: Fix indirect global-variable flow

### PR DESCRIPTION
**Pull request by MathiasVP**
The logic for handling indirections of global variables was all wrong 😭. This PR hopefully fixes this logic 🤞.

The first commit adds a bunch of tests. I've then split the fix into a sequence of small commits. Finally, the last commit accepts the test changes.

There are a couple of new inconsistencies regarding missing `toString` on some esoteric syntax. I haven't investigated those yet, but I don't think they should block this PR.

@jketema I'm requesting a review from you because you initially identified these missing steps. Let me know what I can do to help the review process here!

### CODE REVIEW

**Comment by MathiasVP** [comment]
Here's my investigations so far:

`cpp/unbounded-write` on vim looks good. The new results are caused by us recognizing flow from the output argument: https://github.com/vim/vim/blob/a5a1ec1826c0e43d0282ba4d35c155a97bab3e27/src/option.c#L2016

to this strncpy that stores to a global: https://github.com/vim/vim/blob/a5a1ec1826c0e43d0282ba4d35c155a97bab3e27/src/option.c#L2088

and then flows to another use of the global here: https://github.com/vim/vim/blob/a5a1ec1826c0e43d0282ba4d35c155a97bab3e27/src/memline.c#L2568

**Comment by jketema** [comment]
* https://github.com/github/codeql/pull/14736/commits/bb5a78d3f1ac08c44bf77b8d292a86febbdd52cb also seem to do a replacement at the very bottom, not just a refactoring. What's the purpose of that replacement?
* https://github.com/github/codeql/pull/14736/commits/fd26ae18bf52e2b384add3d98a88532e5fab0f34 seems to do something different than its commit message advertises?

**Comment by MathiasVP** [comment]
> * [bb5a78d](https://github.com/github/codeql/commit/bb5a78d3f1ac08c44bf77b8d292a86febbdd52cb) also seem to do a replacement at the very bottom, not just a refactoring. What's the purpose of that replacement?

Oops. You're right. The bottom line in that commit should actually have been part of https://github.com/github/codeql/commit/fd26ae18bf52e2b384add3d98a88532e5fab0f34.

> * https://github.com/github/codeql/commit/fd26ae18bf52e2b384add3d98a88532e5fab0f34 seems to do something different than its commit message advertises?

I think the commit message is accurate. The commit changes the implementation of `getSourceVariable` for the `GlobalUse` class so that it obtains the SSA variable (which can be thought of as a regular variable along with an indirection count) by calculating the indirection count using the `getIndirection` member predicate (instead of the indirection index).

For consistency this commit also implements a `getIndirection` member predicate for `GlobalDef`s and uses that to implement `getSourceVariable` the same way as `GlobalUse` does it, but since `getIndirection()` just returns the indirection index this doesn't actually change anything in this case.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll [hunk 1]
```diff
@@ -143,11 +143,10 @@ private newtype TDefOrUseImpl =
 private predicate isGlobalUse(
   GlobalLikeVariable v, IRFunction f, int indirection, int indirectionIndex
 ) {
-  exists(VariableAddressInstruction vai, int defIndex |
+  exists(VariableAddressInstruction vai |
     vai.getEnclosingIRFunction() = f and
     vai.getAstVariable() = v and
-    isDef(_, _, _, vai, indirection, defIndex) and
-    indirectionIndex = [0 .. defIndex] + 1
+    isDef(_, _, _, vai, indirection, indirectionIndex)
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll]: The change seems to offset the final argument of `isDef` by 1, is that correct?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll]: Correct. So after this change any write to a global variable with indirection index `indirectionIndex` generates a `GlobalUse` SSA read at the end of the enclosing function with indirection index `indirectionIndex`.

**Review by jketema — APPROVED** [review]
LGTM. Note that I had some 1:1 discussion with @MathiasVP, which clarified most.

**Pull request merged**